pkg/diagnostics/README.md

```diff
@@ -0,0 +1,113 @@
+
+by itself will look for master and node configs (in addition to client
+config file) in the standard locations and use them if found; so this
+should make the Enterprise use case as simple as possible. It's also
+very easy to use configuration files when they are not in the expected
+Enterprise locations:
+
+    openshift ex diagnostics --master-config=... --node-config=...
+
+Having logs in journald is necessary for the current log analysis
+logic. Other usage may have logs going into files, output to stdout,
+combined node/master... it may not be too hard to extend analysis to
+other log sources but the priority has been to look at journald logs
+as created by components in Enterprise deployments (including docker,
+openvswitch, etc.).
+
+Client environment
+==================
+
+The user may only have access as an ordinary user, as a cluster-admin
+user, and/or may be running on a host where OpenShift master or node
+services are operating. The diagnostics will attempt to use as much
+access as the user has available.
+
+A client with ordinary access should be able to diagnose its connection
+to the master and look for problems in builds and deployments for the
+current context.
+
+A client with cluster-admin access should be able to diagnose the
+status of infrastructure.
+
+Writing diagnostics
+===================
+
+Developers are encouraged to add to the available diagnostics as they
+encounter problems that are not easily communicated in the normal
+operations of the program, for example components with misconfigured
+connections, problems that are buried in logs, etc. The sanity you
+save may be your own.
+
+A diagnostic is an object that conforms to the Diagnostic interface
+(see pkg/diagnostics/types/diagnostic.go). The diagnostic object should
+be built in one of the builders in the pkg/cmd/experimental/diagnostics
+package (based on whether it depends on client, cluster-admin, or host
+configuration). When executed, the diagnostic logs its findings into
+a result object. It should be assumed that they may run in parallel.
+
+Diagnostics should prefer providing information over perfect accuracy,
+as they are the first line of (self-)support for users. On the other
+hand, judgment should be exercised to prevent sending users down useless
+paths or flooding them with non-issues that obscure real problems.
+
+* Errors should be reserved for things that are almost certainly broken
+  or causing problems, for example a broken URL.
+* Warnings indicate issues that may be a problem but could be valid for
+  some configurations / situations, for example a node being disabled.
+
+Enabling automation
+===================
+
+Diagnostic messages are designed to be logged either for human consumption
+("text" format) or for scripting/automation ("yaml" or "json" formats). So
+messages should:
+
+* Have an ID that is unique and unchanging, such that automated alerts
+  could filter on specific IDs rather than rely on message text or level.
```

Comment:
I'd be fine with pure IDs (E012345) as long as we ensured uniqueness, or
more readable ones. Halfway between is uncomfortable.



On Aug 30, 2015, at 8:24 PM, Luke Meyer <notifications@github.com> wrote:

In pkg/diagnostics/README.md
<https://github.com/openshift/origin/pull/2049#discussion_r38278279>:

> +paths or flooding them with non-issues that obscure real problems.
> +
> +* Errors should be reserved for things that are almost certainly broken
> +  or causing problems, for example a broken URL.
> +* Warnings indicate issues that may be a problem but could be valid for
> +  some configurations / situations, for example a node being disabled.
> +
> +Enabling automation
> +===================
> +
> +Diagnostic messages are designed to be logged either for human consumption
> +("text" format) or for scripting/automation ("yaml" or "json" formats). So
> +messages should:
> +
> +* Have an ID that is unique and unchanging, such that automated alerts
> +  could filter on specific IDs rather than rely on message text or level.

Because they'd be long, and someone has to type them in the source, and
they show up in the output? And I can't hit '*' in vim to check for
uniqueness? *shrug* They're only semi-readable to make it easier to come up
with unique ones. They could all be E12351. I didn't have any bright ideas
for ensuring uniqueness, until we I18Nize everything and they're message
lookups. Without machine-directed output they're less interesting. If you
feel they should be more readable I can do that.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/2049/files#r38278279>.